src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,157 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/global.h"
+#include "base/logger.h"
+
+BaseLogModel::BaseLogModel(QObject *object)
+    : QAbstractListModel(object)
+    , m_messages(MAX_LOG_MESSAGES)
+{
+}
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    return m_messages.size();
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid())
+        return {};
+
+    const int msgIndex = m_messages.size() - index.row() - 1;
+    if ((msgIndex < 0) || (msgIndex >= m_messages.size()))
+        return {};
+
+    const Message &message = m_messages[msgIndex];
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole;
+    case Qt::UserRole:
+        return message.userRole;
+    case TimeRole:
+        return message.time;
+    case MessageRole:
+        return message.message;
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message& message)
+{
+    // if row is inserted on filled up buffer, the size will not change but because of calling of beginInsertRows function
+    // we'll have ghost rows.
+    if (m_messages.size() == MAX_LOG_MESSAGES) {
+        const int lastMessage = m_messages.size() - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        m_messages.pop_front();
+        endRemoveRows();
+    }
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    m_messages.push_back(message);
+    endInsertRows();
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_messages.clear();
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Msg &msg : asConst(Logger::instance()->getMessages()))
+        handleNewMessage(msg);
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage, Qt::QueuedConnection);
```

Comment:
Why do you use QueuedConnection?

Reply:
the `newLogMessage` is emitted from `addMessage` which can be called from different threads, so I guess this way we don't need any synchronization.

Reply:
It would be nice to add some comment in code.

Reply:
Really signal receiver should know nothing about emitter internals!
You should use `Qt::QueuedConnection` directly only in case you want `QueuedConnection` explicitly for some reason. In this case you can safetely use default `Qt::AutoConnection`. Doc says about it:
>If the receiver lives in the thread that emits the signal, `Qt::DirectConnection` is used. Otherwise, `Qt::QueuedConnection` is used. The connection type is determined when the signal is emitted.

Anyway you should remove the comment above because it creates the false impression that we have an incorrect application design.

Reply:
AFAIK this automatic detection will only work when we use `QThread`